crates/handlers/src/oauth2/end_session.rs

```diff
@@ -0,0 +1,211 @@
+
+    #[error("bad request")]
+    BadRequest,
+
+    #[error("client not found")]
+    ClientNotFound,
+
+    #[error("client is unauthorized")]
+    UnauthorizedClient,
+
+    // #[error("unsupported token type")]
+    // UnsupportedTokenType,
+    #[error("unknown token")]
+    UnknownToken,
+}
+
+impl_from_error_for_route!(mas_storage::RepositoryError);
+
+impl IntoResponse for RouteError {
+    fn into_response(self) -> Response {
+        let sentry_event_id = record_error!(self, Self::Internal(_));
+        let response = match self {
+            Self::Internal(_) => (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                Json(ClientError::from(ClientErrorCode::ServerError)),
+            )
+                .into_response(),
+
+            Self::BadRequest => (
+                StatusCode::BAD_REQUEST,
+                Json(ClientError::from(ClientErrorCode::InvalidRequest)),
+            )
+                .into_response(),
+
+            Self::ClientNotFound => (
+                StatusCode::UNAUTHORIZED,
+                Json(ClientError::from(ClientErrorCode::InvalidClient)),
+            )
+                .into_response(),
+
+            // Self::ClientNotAllowed |
+            Self::UnauthorizedClient => (
+                StatusCode::UNAUTHORIZED,
+                Json(ClientError::from(ClientErrorCode::UnauthorizedClient)),
+            )
+                .into_response(),
+
+            // Self::UnsupportedTokenType => (
+            //     StatusCode::BAD_REQUEST,
+            //     Json(ClientError::from(ClientErrorCode::UnsupportedTokenType)),
+            // )
+            //     .into_response(),
+
+            // If the token is unknown, we still return a 200 OK response.
+            Self::UnknownToken => StatusCode::OK.into_response(),
+        };
+
+        (sentry_event_id, response).into_response()
+    }
+}
+
+impl From<IdTokenError> for RouteError {
+    fn from(_e: IdTokenError) -> Self {
+        Self::UnknownToken
+    }
+}
+
+#[tracing::instrument(name = "handlers.oauth2.end_session.get", skip_all)]
+pub(crate) async fn get(
+    mut rng: BoxRng,
+    clock: BoxClock,
+    State(key_store): State<Keystore>,
+    State(url_builder): State<UrlBuilder>,
+    mut repo: BoxRepository,
+    activity_tracker: BoundActivityTracker,
+    Query(params): Query<EndSessionParam>,
+    cookie_jar: CookieJar,
+) -> Result<Response, RouteError> {
+    let (session_info, cookie_jar) = cookie_jar.session_info();
+
+    let browser_session_id = session_info
+        .current_session_id()
+        .ok_or(RouteError::BadRequest)?;
+
+    let browser_session = repo
+        .browser_session()
+        .lookup(browser_session_id)
+        .await?
+        .ok_or(RouteError::BadRequest)?;
+
+    let oauth_session = repo
+        .oauth2_session()
+        .find_by_browser_session(browser_session.id)
+        .await?
+        .ok_or(RouteError::BadRequest)?;
```

Comment:
You can have many OAuth session started from a single browser session, so this way is not the right approach here.

Instead, best is to decode the JWT (but not necessarily verify it, maybe `verify_id_token` should be changed a little bit?) to get the claims we need and go from there? I'm unclear what we should be checking exactly on that JWT though, I don't know whether there is something that would let us link it to the current browser session or not 🤔